### ISSUE TITLE
S3 PutObject fails when Content-Type is specified

### ISSUE THREAD

**Issue by jluszcz** [awslabs/aws-sdk-rust#131]
### Version

```
├── aws-sdk-s3 v0.0.9-alpha (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53)
│   ├── aws-auth v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53)
│   │   ├── smithy-http v0.0.1 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53)
│   │   │   ├── smithy-types v0.0.1 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53)
│   ├── aws-endpoint v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53)
│   │   ├── aws-types v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53)
│   │   └── smithy-http v0.0.1 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   ├── aws-http v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53)
│   │   ├── aws-types v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   │   ├── smithy-http v0.0.1 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   │   ├── smithy-types v0.0.1 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   ├── aws-hyper v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53)
│   │   ├── aws-auth v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   │   ├── aws-endpoint v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   │   ├── aws-http v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   │   ├── aws-sig-auth v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53)
│   │   │   ├── aws-auth v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   │   │   ├── aws-types v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   │   │   ├── smithy-http v0.0.1 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   │   ├── protocol-test-helpers v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53)
│   │   ├── smithy-client v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53)
│   │   │   ├── protocol-test-helpers v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   │   │   ├── smithy-http v0.0.1 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   │   │   ├── smithy-http-tower v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53)
│   │   │   │   ├── smithy-http v0.0.1 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   │   │   ├── smithy-types v0.0.1 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   │   ├── smithy-http v0.0.1 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   │   ├── smithy-http-tower v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   │   ├── smithy-types v0.0.1 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   ├── aws-sig-auth v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   ├── aws-types v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   ├── smithy-http v0.0.1 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   ├── smithy-types v0.0.1 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53) (*)
│   └── smithy-xml v0.1.0 (https://github.com/awslabs/aws-sdk-rust?tag=v0.0.9-alpha#52164d53)
```

### Platform

```
Darwin MacBook-Air.local 20.5.0 Darwin Kernel Version 20.5.0: Sat May  8 05:10:31 PDT 2021; root:xnu-7195.121.3~9/RELEASE_ARM64_T8101 arm64
```

This also occurred on Amazon Linux 2.

### AWS Services

S3

### Description

Specifying content-type on S3 PutObject requests fails with `SignatureDoesNotMatch`

Reproduction:

```
#[tokio::main]
async fn main() -> Result<()> {
    let s3_client = s3::Client::from_env();

    let content = r#"<html>
    <head>
    <title>Title</title>
    </head>
    <body>
    <p>Paragraph</p>
    </body>
    </html>"#;

    let request = s3_client
        .put_object()
        .bucket(BUCKET)
        .key("test-1")
        .body(ByteStream::from(Vec::from(content.as_bytes())));
    request.send().await?;

    let request = s3_client
        .put_object()
        .bucket(BUCKET)
        .key("test-2")
        .content_type("text/html")
        .body(ByteStream::from(Vec::from(content.as_bytes())));
    request.send().await.expect("Failed w/ Content-Type");

    Ok(())
}
```

```
thread 'main' panicked at 'Failed w/ Content-Type: ServiceError { err: PutObjectError { kind: Unhandled(Error { code: Some("SignatureDoesNotMatch"), message: Some("The request signature we calculated does not match the signature you provided. Check your key and signing method."),
```

I expected to see this happen: `PutObject` succeeds with content-type on the object set to `text/html`

Instead, this happened: `SignatureDoesNotMatch` error

**Comment by brainstorm** [comment]
I've observed that passing a not correct key object (i.e accidentally passing `{{ closure }}` string) string also triggers the unhandled error... it should filter according to the official rules/regexes:

https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html

And return an appropriate error, I reckon?

**Comment by jdisanti** [comment]
This was fixed in [`v0.0.12-alpha`](https://github.com/awslabs/aws-sdk-rust/releases/tag/v0.0.12-alpha). Thanks for the PR, @eagletmt!

@brainstorm - The issue you're talking about sounds vaguely familiar. I think we might have fixed that by fixing URL escaping. But if you still see it in the latest version, please file a separate issue. Thanks!
